### DISCUSSION TITLE
[gdal] How to build gdal plugins/drivers?

### DISCUSSION THREAD

**Discussion by michaeladamkatz** [microsoft/vcpkg#32438] [Q&A]
I built gdal successfully, but the only thing in lib/gdalplugins is an .ini file. I would like to build drivers for as many raster formats as possible. Does vcpkg have the ability to build the GDAL driver .dll files?

**Comment by autoantwort** [comment]
There are a lot of features you can enable: https://vcpkg.link/ports/gdal

**Comment by dg0yt** [comment]
Drivers can be enabled or disabled via port features. The port tries to avoid uncontrolled dependencies.
But we don't make them plugins.
You may pass additional [per-port](https://learn.microsoft.com/en-us/vcpkg/users/triplets#per-port-customization) cmake options from a custom triplet via [`VCPKG_CMAKE_CONFIGURE_OPTIONS`](https://learn.microsoft.com/en-us/vcpkg/users/triplets#vcpkg_cmake_configure_options),

**Comment by michaeladamkatz** [comment]
Kai, and autoantwort, thanks but can you point me to the specific feature or features that would turn on driver creation? I did look at the various config files and features but they all seemed to be about other libraries like using curl or sqlite3, not about plugins/drivers for file formats. As two examples, I want to be able to open .ecw and .sid files and I didn't see any settings to turn those on specifically or to turn on all file formats.
Also, Kai, when you say "we don't make them plugins" I can't tell if you mean that the drivers are linked statically into the gdal.dll, or what...


    On Friday, July 7, 2023 at 09:42:01 AM PDT, Kai Pastor ***@***.***> wrote:  
 
 


Drivers can be enabled or disabled via port features. But the port tries to avoid uncontrolled dependencies.
But we don't make them plugins.
You may pass additional per-port cmake options from a custom triplet via VCPKG_CMAKE_CONFIGURE_OPTIONS,

**Reply by dg0yt** [reply]
The drivers are in the single gdal lib. 
And yes, most features are named like the dependencies, however it maps closely to gdal's drivers.

Some formats like .ecw need external SDKs which aren't (and cannot be) available in vcpkg. So they can't be tested in vcpkg CI. It may be possible to add such features, but they would be more prone to regressions.

**Reply by michaeladamkatz** [reply]
Thanks. I have moved my next question to https://gis.stackexchange.com/questions/463025/reading-mrsid-sid-files-with-a-gdal-built-by-vcpkg. If you know those specifics, please let me know. If the specifics have already been implicitly answered by your previous responses, I apologize for not understanding.

**Reply by dg0yt** [reply]
The answer was given indeed:
You have to provide additional options to vcpkg at build time, https://github.com/microsoft/vcpkg/discussions/32438#discussioncomment-6386560.

The whole set of sources and options is used to identify binary artifacts for caching in vcpkg, so it is mandatory that the build doesn't use anything just because "it is there" but because "it is explicit".

Relevant cmake options are documented at the gdal side:
MRSID: https://gdal.org/development/building_from_source.html#mrsid
ECW: https://gdal.org/development/building_from_source.html#ecw

**Reply by dg0yt** [reply]
And remember that for windows, it may be more complicated, due to different runtimes for release vs. debug.

**Reply by michaeladamkatz** [reply]
Thanks very much for the detail. I see now that my confusion was that I was thinking of vcpkg-mediated build as sort of its own world and the options for the manual build as a separate world, but of course the vcpkg build runs the manual build, and needs to pass through all of its options.

**Comment by michaeladamkatz** [comment]
In case it helps others, here are the steps I took to use MrSID and ECW through GDAL:

(1a) Download the 32 bit Visual Studio 2022 DSDK from https://www.extensis.com/support/developers. Unzip it to C:\Users\michael.katz\Documents\vcpkg\sdk\MrSID_DSDK-9.5.5.5244-win32-vc17.

(1b) Go to https://supportsi.hexagon.com/help/s/article/ERDAS-ECW-JP2-SDK-Read-Only-Redistributable-download and download the appropriate SDK in the top-right corner. Unzip the download to C:\Users\michael.katz\Documents\vcpkg\sdk\ecw\Hexagon\ERDAS_ECW_JPEG_2000_SDK_5.5.0\Desktop_Read-Only.

(2) Edit C:\\Users\\michael.katz\\Documents\\vcpkg\\triplets\\x86-windows.cmake. (You can see the meaning of the flags at https://gdal.org/development/building_from_source.html.)

```
set(VCPKG_TARGET_ARCHITECTURE x86)
set(VCPKG_CRT_LINKAGE static)
set(VCPKG_LIBRARY_LINKAGE static)
set(VCPKG_BUILD_TYPE release)

# CMAKE_CURRENT_LIST_DIR is the triplets directory
# such as C:/Users/michael.katz/Documents/vcpkg/triplets
message( STATUS "\n\n-----------------------CMAKE_CURRENT_LIST_DIR = ${CMAKE_CURRENT_LIST_DIR}\n\n" )

if (PORT MATCHES "gdal")
	string(
		CONCAT x
		" \"-DMRSID_INCLUDE_DIR=${CMAKE_CURRENT_LIST_DIR}/../sdk/MrSID_DSDK-9.5.5.5244-win32-vc17/Raster_DSDK/include\""
		" \"-DMRSID_LIBRARY=${CMAKE_CURRENT_LIST_DIR}/../sdk/MrSID_DSDK-9.5.5.5244-win32-vc17/Raster_DSDK/lib/lti_dsdk.lib\""
		" \"-DGDAL_USE_MRSID=ON\""
		" \"-DECW_INCLUDE_DIR=${CMAKE_CURRENT_LIST_DIR}/../sdk/ecw/Hexagon/ERDAS_ECW_JPEG_2000_SDK_5.5.0/Desktop_Read-Only/include\""
		" \"-DECW_LIBRARY=${CMAKE_CURRENT_LIST_DIR}/../sdk/ecw/Hexagon/ERDAS_ECW_JPEG_2000_SDK_5.5.0/Desktop_Read-Only/lib/vc141/Win32/NCSEcw.lib\""
		" \"-DGDAL_USE_ECW=ON"
	)
	set(VCPKG_CMAKE_CONFIGURE_OPTIONS "${VCPKG_CMAKE_CONFIGURE_OPTIONS} ${x}" )
endif()

message( STATUS "\n\n-----------------------VCPKG_CMAKE_CONFIGURE_OPTIONS = ${VCPKG_CMAKE_CONFIGURE_OPTIONS}\n\n" )

message( STATUS "\n\n-----------------------FEATURES = ${FEATURES}\n\n" )

# message( STATUS "\n\n-----------------------VCPKG_CXX_FLAGS = ${VCPKG_CXX_FLAGS}\n\n" )
```

(3) Build gdal, but first remove the old gdal:

```
PS C:\Users\michael.katz\Documents\vcpkg> rmdir C:\Users\michael.katz.la\AppData\Local\vcpkg\archives

PS C:\Users\michael.katz\Documents\vcpkg> .\vcpkg.exe remove gdal

PS C:\Users\michael.katz\Documents\vcpkg> rm C:\Users\michael.katz\Documents\vcpkg\buildtrees\gdal\x86-windows-rel\CMakeCache.txt

PS C:\Users\michael.katz\Documents\vcpkg> .\vcpkg.exe install gdal
```

(4a) Add `...MrSID_DSDK-9.5.5.5244-win32-vc17\Raster_DSDK\lib\lti_dsdk.lib` as a linker input for your project.

(4b) Add `...Desktop_Read-Only\lib\vc141\Win32\NCSEcw.lib` as a linker input for your project.

(5a) Copy `...MrSID_DSDK-9.5.5.5244-win32-vc17\Raster_DSDK\lib\lti_dsdk_9.5.dll` to the Release directory of your project, so it's a sibling to your `app.exe`.

(5b) Copy `...Desktop_Read-Only\redistributable\vc141\Win32\NCSEcw.dll` to the Release directory of your project, so it's a sibling to your `app.exe`.

(6) Rebuild your project.

**Reply by dg0yt** [reply]
> ~~~cmake
> set(ENV{CMAKE_WINDOWS_KITS_10_DIR} "C:\\Program Files (x86)\\Windows Kits\\10")
> set(VCPKG_CMAKE_CONFIGURE_OPTIONS "-DCMAKE_WINDOWS_KITS_10_DIR=C:/Program Files (x86)/Windows Kits/10\"")
> set(VCPKG_ENV_PASSTHROUGH CMAKE_WINDOWS_KITS_10_DIR)

Is this really needed? vcpkg should handle standard MSVC toolchain stuff.

> ~~~cmake
> 	string(
> 		CONCAT x
> 		" \"-DMRSID_INCLUDE_DIR=${CMAKE_CURRENT_LIST_DIR}/../sdk/MrSID_DSDK-9.5.5.5244-win32-vc17/Raster_DSDK/include\""
> 		" \"-DMRSID_LIBRARY=${CMAKE_CURRENT_LIST_DIR}/../sdk/MrSID_DSDK-9.5.5.5244-win32-vc17/Raster_DSDK/lib/lti_dsdk.lib\""
> 		" \"-DGDAL_USE_MRSID=ON\""
> 		" \"-DECW_INCLUDE_DIR=${CMAKE_CURRENT_LIST_DIR}/../sdk/ecw/Hexagon/ERDAS_ECW_JPEG_2000_SDK_5.5.0/Desktop_Read-Only/include\""
> 		" \"-DECW_LIBRARY=${CMAKE_CURRENT_LIST_DIR}/../sdk/ecw/Hexagon/ERDAS_ECW_JPEG_2000_SDK_5.5.0/Desktop_Read-Only/lib/vc141/Win32/NCSEcw.lib\""
> 		" \"-DGDAL_USE_ECW=ON"
> 	)
> 	set(VCPKG_CMAKE_CONFIGURE_OPTIONS "${VCPKG_CMAKE_CONFIGURE_OPTIONS} ${x}" )

Do not use `string(CONCAT ...)`. These options are expected to form a CMake list.
I would rewrite this as:
~~~cmake
        set(MRSID_SDK "${CMAKE_CURRENT_LIST_DIR}/../sdk/MrSID_DSDK-9.5.5.5244-win32-vc17/Raster_DSDK")
 	list(APPEND VCPKG_CMAKE_CONFIGURE_OPTIONS
 		"-DGDAL_USE_MRSID=ON"
 		"-DMRSID_INCLUDE_DIR=${MRSID_SDK}/include"
 		"-DMRSID_LIBRARY=${MRSID_SDK}/lib/lti_dsdk.lib"
	)
        set(ECW_SDK "${CMAKE_CURRENT_LIST_DIR}/../sdk/ecw/Hexagon/ERDAS_ECW_JPEG_2000_SDK_5.5.0/Desktop_Read-Only")
 	list(APPEND VCPKG_CMAKE_CONFIGURE_OPTIONS
 		"-DGDAL_USE_ECW=ON"
 		"-DECW_INCLUDE_DIR=${ECW_SDK}/include"
 		"-DECW_LIBRARY=${ECW_SDK}/lib/vc141/Win32/NCSEcw.lib"
	)
~~~
(Generally quote for CMake, and let CMake handle the quoting for the invoked commands.)
In case of CRT issues, you might need to per-build-type options:
~~~cmake
 	list(APPEND VCPKG_CMAKE_CONFIGURE_OPTIONS_RELEASE
 		"-DMRSID_LIBRARY=${MRSID_SDK}/lib/lti_dsdk.lib"
 	)  
 	list(APPEND VCPKG_CMAKE_CONFIGURE_OPTIONS_DEBUG
 		"-DMRSID_LIBRARY=${MRSID_SDK}/lib/<...>"
 	) 
	# etc.
~~~

**Reply by michaeladamkatz** [reply]
Thanks for the pointer on lists to get me out of quoting trickery. For the Windows Kits path, I had some trouble with my corporate policy not allowing registry reading. I'm not sure if those attempts to define the Windows Kits path will stay in, but for now they're working, and at least not hurting.

**Reply by dg0yt** [reply]
> For the Windows Kits path, I had some trouble with my corporate policy not allowing registry reading.

So this is unrelated to gdal, and it has side effects, but readers might not realize it. You can keep it on your disk, but IMO it is not good to have it as part of the recipe here.

**Reply by michaeladamkatz** [reply]
Agreed. I have taken it out of the code above.
